hiro/cocoa/widget/hex-edit.cpp

```diff
@@ -5,43 +5,163 @@
+      [tableView setDataSource:self];
+      [tableView setHeaderView:nil];
+      
+      // remove padding and stuff to fit all 16 columns
+      [tableView setIntercellSpacing:NSMakeSize(0, 0)];
+      [tableView setStyle:NSTableViewStylePlain];
+      NSTableColumn* addressCol = [[NSTableColumn alloc] initWithIdentifier:@"Address"];
+      addressCol.editable = NO;
+      [tableView addTableColumn:addressCol];
+      
+      for (int i = 0; i < hexEdit->columns(); i++) {
+        NSTableColumn* col = [[NSTableColumn alloc] initWithIdentifier:[NSString stringWithFormat:@"%d",
+                                                                        i]];
+        col.width = 20;
+        col.title = @"";
+        [tableView addTableColumn:col];
+      }
+      /*
+       NSTableColumn* hexCol = [[NSTableColumn alloc] initWithIdentifier:@"Hex"];
+       [hexCol setTitle:@"Hex"];
+       [tableView addTableColumn:hexCol];
+       */
+      
+      NSTableColumn* ansiCol = [[NSTableColumn alloc] initWithIdentifier:@"ANSI"];
+      ansiCol.editable = NO;
+      [tableView addTableColumn:ansiCol];
+      
+      // must programmatically do this so the table shows up.
+      [self setDocumentView:tableView];
+      [self setHasVerticalScroller:YES];
+    }
   }
   return self;
 }
 
-@end
-
-namespace hiro {
-
-auto pHexEdit::construct() -> void {
-  cocoaView = cocoaHexEdit = [[CocoaHexEdit alloc] initWith:self()];
-}
-
-auto pHexEdit::destruct() -> void {
-  [cocoaView removeFromSuperview];
+- (NSTableView *) tableView {
+  return tableView;
 }
 
-auto pHexEdit::setAddress(u32 offset) -> void {
+- (hiro::mHexEdit *) hexEdit {
+  return hexEdit;
 }
 
-auto pHexEdit::setBackgroundColor(Color color) -> void {
-}
-
-auto pHexEdit::setColumns(u32 columns) -> void {
-}
 
-auto pHexEdit::setForegroundColor(Color color) -> void {
+- (NSInteger) numberOfRowsInTableView:(NSTableView *)tableView {
+  return (max(1, hexEdit->length()) + hexEdit->columns() - 1) / hexEdit->columns();
 }
 
-auto pHexEdit::setLength(u32 length) -> void {
+-(id) tableView:(NSTableView *) tableView
+objectValueForTableColumn:(NSTableColumn *) tableColumn
+            row:(NSInteger) row {
+  // return a string with the content of (row, column)
+  // u32 address = hexEdit->address() + row * hexEdit->columns();
+  u32 address = row * hexEdit->columns();
+  // address only
+  if ([[tableColumn identifier] isEqualToString:@"Address"]) {
+    return [NSString stringWithUTF8String:hex(address, 8L).data()];
+  } else if ([[tableColumn identifier] isEqualToString:@"ANSI"]){
+    // create the ansi string by concatenating
+    NSMutableString *output = [NSMutableString stringWithCapacity:hexEdit->columns()];
+    for (auto column : range(hexEdit->columns())) {
+      if (address < hexEdit->length()) {
+        u8 data = hexEdit->doRead(address++);
+        [output appendString:[NSString stringWithFormat:@"%c",
+                              (data >= 0x20 && data <= 0x7e ? (char)data : '.')]];
```

Comment:
This is something of a nit/bikeshedding, since the terminology isn't exposed to the user currently, but "ANSI" is sort of a Windows-ism, which just sort of generically means "the system's default encoding". On macOS, that term isn't really used, since basically everything is UTF-8. It looks like when we're interpolating the `char` here, we're using the UTF-8 system-default interpolation of the character, but just constrained to the range 32-127 of printable characters that have only one codepoint. Not worth changing the names of everything (unless you feel like it), but if the column header is ever exposed to the user, worth keeping in mind. I also personally would prefer a whitespace instead of a `.`, since the actual `.` character is within the 32-127 range, but others might have other opinions.